### PR TITLE
Optimize scheduler latches to avoid latch-wait

### PR DESCRIPTION
**Pull request by Little-Wallace**
###  What have you changed?

The latches scheduler would prevent key writing ,if there has been one key in waiting queue. I changed this behavior to check the waiting if  there has been a  hash value repeat with it .

I set `scheduler-concurrency` to `512 * 1024` by default. If  there are 1000K keys written at the same time,  every key would be compared in the deque only twice. I think it is efficient enough.

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

###  Benchmark
```
config:
      3 TIKV
      tikv.yml:
           apply-pool-size: 4
script:
      sysbench oltp_update_non_index --config-file=config --threads=16 --tables=16 --table_size=4000000 prepare
result:
      master: 11min 16s
      this PR:  9min 17s
```

### CODE REVIEW

**Comment by siddontang** [comment]
seem here we need to construct a case that contains many fake key hash conflicts, right?

**Comment by Little-Wallace** [comment]
@siddontang  Yes, I would add a case and use a mock Hasher.

File: src/storage/config.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ pub const DEFAULT_DATA_DIR: &str = "./";
 pub const DEFAULT_ROCKSDB_SUB_DIR: &str = "db";
 const DEFAULT_GC_RATIO_THRESHOLD: f64 = 1.1;
 const DEFAULT_MAX_KEY_SIZE: usize = 4 * 1024;
-const DEFAULT_SCHED_CONCURRENCY: usize = 2048000;
+const DEFAULT_SCHED_CONCURRENCY: usize = 1024 * 512;
```
Comment (zhangjinpeng1987) [inline src/storage/config.rs]: Please limit the size in config, because some existing cluster already used a 100x size, it will cause OOM when apply to this new mechanism.

**Comment by sticnarf** [comment]
It seems very similar to a concurrent hash map now. Why not use a concurrent hash map library?

File: src/storage/config.rs [hunk 2]
```diff
@@ -58,6 +59,12 @@ impl Config {
         if self.data_dir != DEFAULT_DATA_DIR {
             self.data_dir = config::canonicalize_path(&self.data_dir)?
         }
+        if self.scheduler_concurrency > MAX_SCHED_CONCURRENCY {
+            warn!("TiKV optimize latch in v4.0, so it is not necessary use large schedule concurrency. \
```
Comment (zhangjinpeng1987) [inline src/storage/config.rs]: TiKV had optimized

**Comment by zhangjinpeng1987** [comment]
> It seems very similar to a concurrent hash map now. Why not use a concurrent hash map library?

In that case we must clac the hash twice?

File: src/storage/txn/latch.rs [hunk 3]
```diff
@@ -29,6 +30,49 @@ impl Latch {
...
+            }
+        }
+        None
+    }
+
+    pub fn wait_for_wake(&mut self, s: usize, cid: u64) {
+        self.waiting.push_back(Some((s, cid)));
+    }
+
+    pub fn maybe_shrink(&mut self) {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: Can be merged into pop_front.

**Comment by sticnarf** [comment]
> > It seems very similar to a concurrent hash map now. Why not use a concurrent hash map library?
> 
> In that case we must clac the hash twice?

Let's consider `Latches` just as a `ConcurrentHashMap<Key, VecDeque<Task>>`, where does the second hash occur?

**Comment by Little-Wallace** [comment]
@sticnarf  It means that we would store every key in `HashMap`.

File: src/storage/txn/latch.rs [hunk 4]
```diff
@@ -19,7 +20,7 @@ const WAITING_LIST_SHRINK_SIZE: usize = 32;
 #[derive(Clone)]
 struct Latch {
     // store waiting commands
-    pub waiting: VecDeque<u64>,
+    pub waiting: VecDeque<Option<(usize, u64)>>,
```
Comment (AndreMouche) [inline src/storage/txn/latch.rs]: More memory?

File: src/storage/txn/latch.rs [hunk 5]
```diff
@@ -29,6 +30,49 @@ impl Latch {
...
+    }
+
+    pub fn maybe_shrink(&mut self) {
+        // Pop item which is none to make queue not too long.
+        while !self.waiting.is_empty() {
+            let item = self.waiting.front().unwrap();
+            if item.is_none() {
+                self.waiting.pop_front();
+            } else {
+                break;
```
Comment (AndreMouche) [inline src/storage/txn/latch.rs]: too complexed? How about the performance with other solutions like make Latches to a Concurrency hash map?

Comment (Connor1996) [reply src/storage/txn/latch.rs]: +1

File: src/storage/txn/latch.rs (hunk 4 above)
Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: Just three times memory per Latch. But we could allocate less capacity of Latch. So it cost less memory in general.

Comment (sticnarf) [reply src/storage/txn/latch.rs]: I think it's three times per task instead of per latch. It's hard to say whether the memory usage is reduced.

File: src/storage/txn/latch.rs (hunk 5 above)
Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: No. ConcurrencyHashmap could only support simple operation such as `get` and `put`, but here we need a queue could store request id which has the same key.

Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: If you use ConcurrencyHashMap, you still need write such code:
```
latches: ConcurrencyHashMap<Key, Mutex<VecDeque<u64>>;
```
Obviously, `ConcurrencyHashMap` would perform much worse than array.

Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: If you want a better performance than this PR, you maybe need a ConcurrencyHashMap and ConcurrencyQueue. I think it would be more complex to deal with write-request in different thread.

Comment (sticnarf) [reply src/storage/txn/latch.rs]: How do you think it compares to `ConcurrentHashMap<ArcKey, Task>`? That's one more allocation but the benefits is simpler latch, less memory usage and low initialization cost.

File: src/storage/txn/latch.rs (hunk 4 above)
Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: No. In product environment, user usually use a much larger scheduler-concurrency to avoid high latch-wait latency. For example,  a bank may set scheduler-concurrency 102400000, which is five times of default value and cost more than 400 MB (pointer in VecDeque also cost memory) even if TiKV does not access any write request. 
According to my implementation, you could only need set scheduler-concurrency  512000,   which is 1/200 of 102400000 and cost no more than 2MB.
If there are 1000k keys ready to write into TiKV server, every key would cost 3 * 8B memory. All of they just cost 24MB memory, which is much less master.
So this PR would save much memory.

File: src/storage/txn/latch.rs (hunk 5 above)
Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: But how do you deal with tasks which has the same key?  Don't you need to store them in queue?

Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: Since we must store requests with the same key in the same queue,  queue is necessary.  I just merge bucket of hashtable and queue of request into one structure.

**Comment by Little-Wallace** [comment]
@sticnarf  any more comments?

**Comment by sticnarf** [comment]
I'm fine with the idea.

**Comment by Little-Wallace** [comment]
@AndreMouche

**Comment by Little-Wallace** [comment]
@AndreMouche  addressed comments, PTAL.

File: src/storage/txn/latch.rs [hunk 6]
```diff
@@ -149,9 +194,16 @@ impl Latches {
     where
         H: Hash,
     {
+        // TODO: avoid Hasher construct
         let mut s = DefaultHasher::new();
         key.hash(&mut s);
-        (s.finish() as usize) & (self.size - 1)
```
Comment (sticnarf) [inline src/storage/txn/latch.rs]: Need to update comments which mention "slot ID". Now a lock stores key hashes only instead of slot ID.

File: src/storage/txn/latch.rs [hunk 7]
```diff
@@ -149,9 +194,16 @@ impl Latches {
     where
         H: Hash,
     {
+        // TODO: avoid Hasher construct
```
Comment (sticnarf) [inline src/storage/txn/latch.rs]: Creating a hasher is very cheap. If we really need to reduce the cost, we must change the hash algorithm

File: tests/integrations/server/lock_manager.rs [hunk 8]
```diff
@@ -30,7 +30,12 @@ fn acquire_pessimistic_lock(
     req.for_update_ts = ts;
     req.lock_ttl = 20;
     req.is_first_lock = false;
-    client.kv_pessimistic_lock(&req).unwrap()
+    let mut resp = client.kv_pessimistic_lock(&req).unwrap();
+    if resp.get_region_error().has_stale_command() {
```
Comment (sticnarf) [inline tests/integrations/server/lock_manager.rs]: Shouldn't be included in this PR? Maybe https://github.com/tikv/tikv/pull/6155 solves it.

File: src/storage/txn/latch.rs [hunk 9]
```diff
@@ -29,6 +30,58 @@ impl Latch {
             waiting: VecDeque::new(),
         }
     }
+
+    pub fn front(&self, s: usize) -> Option<u64> {
```
Comment (sticnarf) [inline src/storage/txn/latch.rs]: The method name `front` is hard to understand

File: tests/integrations/server/lock_manager.rs (hunk 8 above)
Comment (youjiali1995) [reply tests/integrations/server/lock_manager.rs]: Yep. I think https://github.com/tikv/tikv/pull/6155 is ready for review.

File: src/storage/txn/latch.rs [hunk 10]
```diff
@@ -126,19 +178,12 @@ impl Latches {
     /// Preconditions: the caller must ensure the command is at the front of the latches.
     pub fn release(&self, lock: &Lock, who: u64) -> Vec<u64> {
         let mut wakeup_list: Vec<u64> = vec![];
-        for i in &lock.required_slots[..lock.owned_count] {
-            let mut latch = self.slots[*i].lock();
-            let front = latch.waiting.pop_front().unwrap();
+        for &i in &lock.required_slots[..lock.owned_count] {
+            let mut latch = self.lock_latch(i);
+            let (_, front) = latch.pop_front(i).unwrap();
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: assert_eq!(slot, i);

File: src/storage/txn/latch.rs [hunk 11]
```diff
@@ -29,6 +30,58 @@ impl Latch {
...
+
+    pub fn wait_for_wake(&mut self, hash_key: usize, cid: u64) {
+        self.waiting.push_back(Some((hash_key, cid)));
+    }
+
+    // For some hot keys, the waiting list maybe very long, so we should shrink the waiting
+    // VecDeque after pop.
+    fn maybe_shrink(&mut self) {
+        // Pop item which is none to make queue not too long.
+        while let Some(item) = self.waiting.pop_front() {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: I think check if the front is Some(None) before pop is better?

File: src/storage/txn/latch.rs (hunk 9 above)
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: Should be private method

File: tests/integrations/server/lock_manager.rs (hunk 8 above)
Comment (Little-Wallace) [reply tests/integrations/server/lock_manager.rs]: Ok, I have removed this code.

File: src/storage/txn/latch.rs [hunk 12]
```diff
@@ -29,12 +30,69 @@ impl Latch {
...
+        }
+    }
 }
 
 /// Lock required for a command.
 #[derive(Clone)]
 pub struct Lock {
-    /// The slot IDs of the latches that a command must acquire before being able to be processed.
+    /// The hash value of the keys that a command must acquire before being able to be processed.
     pub required_slots: Vec<usize>,
```
Comment (sticnarf) [inline src/storage/txn/latch.rs]: some variable names need updating as well.

File: src/storage/txn/latch.rs [hunk 13]
```diff
@@ -5,21 +5,22 @@ use std::collections::VecDeque;
...
+/// Latches are indexed by slot IDs. The keys of a command are hashed into unsigned number,
+/// then the command is added to the waiting queues of the latches.
 ///
 /// If command A is ahead of command B in one latch, it must be ahead of command B in all the
 /// overlapping latches. This is an invariant ensured by the `gen_lock`, `acquire` and `release`.
 #[derive(Clone)]
 struct Latch {
     // store waiting commands
-    pub waiting: VecDeque<u64>,
+    pub waiting: VecDeque<Option<(usize, u64)>>,
```
Comment (youjiali1995) [inline src/storage/txn/latch.rs]: `usize` is hash value and `u64` is command id?

Comment (sticnarf) [reply src/storage/txn/latch.rs]: Yes. Need more comments.

Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: yes

File: src/storage/txn/latch.rs [hunk 14]
```diff
@@ -5,21 +5,22 @@ use std::collections::VecDeque;
...
 
-const WAITING_LIST_SHRINK_SIZE: usize = 32;
+const WAITING_LIST_SHRINK_SIZE: usize = 8;
+const WAITING_LIST_MAX_CAPACITY: usize = 16;
 
 /// Latch which is used to serialize accesses to resources hashed to the same slot.
 ///
-/// Latches are indexed by slot IDs. The keys of a command are hashed to slot IDs, then the command
-/// is added to the waiting queues of the latches.
+/// Latches are indexed by slot IDs. The keys of a command are hashed into unsigned number,
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: s/number/numbers

File: src/storage/txn/latch.rs [hunk 15]
```diff
@@ -126,32 +183,32 @@ impl Latches {
...
     {
         let mut s = DefaultHasher::new();
         key.hash(&mut s);
-        (s.finish() as usize) & (self.size - 1)
+
+        s.finish()
+    }
+
+    #[inline]
+    fn lock_latch(&self, s: u64) -> MutexGuard<Latch> {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: s/s/hash

File: src/storage/txn/latch.rs [hunk 16]
```diff
@@ -29,34 +30,91 @@ impl Latch {
             waiting: VecDeque::new(),
         }
     }
+
+    /// Find the first command ID in the queue whose hash value is equal to s.
+    pub fn get_first_req_by_hash(&self, s: u64) -> Option<u64> {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: s/s/hash

File: src/storage/txn/latch.rs [hunk 17]
```diff
@@ -29,34 +30,91 @@ impl Latch {
...
+        for item in self.waiting.iter() {
+            if let Some((x, y)) = item {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: if let Some((h, cid)) = item {
  is *h == hash {
    return Some(*cid)
  }
}

File: src/storage/txn/latch.rs [hunk 18]
```diff
@@ -29,34 +30,91 @@ impl Latch {
...
+                    if *v == key_hash {
+                        return it.take();
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    pub fn wait_for_wake(&mut self, hash_key: u64, cid: u64) {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: s/hash_key/key_hash

File: src/storage/txn/latch.rs [hunk 19]
```diff
@@ -29,34 +30,91 @@ impl Latch {
...
+    }
 }
 
 /// Lock required for a command.
 #[derive(Clone)]
 pub struct Lock {
-    /// The slot IDs of the latches that a command must acquire before being able to be processed.
-    pub required_slots: Vec<usize>,
+    /// The hash value of the keys that a command must acquire before being able to be processed.
+    pub required_keys: Vec<u64>,
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: s/required_keys/required_hashes

File: src/storage/txn/latch.rs [hunk 20]
```diff
@@ -5,21 +5,22 @@ use std::collections::VecDeque;
...
+/// Latches are indexed by slot IDs. The keys of a command are hashed into unsigned numbers,
+/// then the command is added to the waiting queues of the latches.
 ///
 /// If command A is ahead of command B in one latch, it must be ahead of command B in all the
 /// overlapping latches. This is an invariant ensured by the `gen_lock`, `acquire` and `release`.
 #[derive(Clone)]
 struct Latch {
-    // store waiting commands
-    pub waiting: VecDeque<u64>,
+    // store hash value of the key and command ID which required this key.
```
Comment (sticnarf) [inline src/storage/txn/latch.rs]: Could you please also comment what a `None` means? It will help people understand the code here.

File: src/storage/txn/latch.rs [hunk 21]
```diff
@@ -86,33 +144,32 @@ impl Latches {
         H: Hash,
     {
         // prevent from deadlock, so we sort and deduplicate the index
-        let mut slots: Vec<usize> = keys.iter().map(|x| self.calc_slot(x)).collect();
-        slots.sort();
-        slots.dedup();
-        Lock::new(slots)
+        let mut hashes: Vec<u64> = keys.iter().map(|x| self.calc_slot(x)).collect();
+        hashes.sort();
```
Comment (sticnarf) [inline src/storage/txn/latch.rs]: Sorting the hashes cannot prevent deadlock. For example, we have 1024 slots. Command A has key hashes `[0, 1025]` while command B has key hashes `[1, 1024]`. Then there can be a deadlock.
So here I believe you need to sort by slot id.

File: src/storage/txn/latch.rs [hunk 22]
```diff
@@ -29,34 +30,91 @@ impl Latch {
...
+    pub required_hashes: Vec<u64>,
 
     /// The number of latches that the command has acquired.
     pub owned_count: usize,
 }
 
 impl Lock {
     /// Creates a lock.
-    pub fn new(required_slots: Vec<usize>) -> Lock {
+    pub fn new(required_keys: Vec<u64>) -> Lock {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: s/required_keys/required_hashes

File: src/storage/txn/latch.rs (hunk 21 above)
Comment (Little-Wallace) [reply src/storage/txn/latch.rs]: Good. This is a bug.

Comment (zhangjinpeng1987) [reply src/storage/txn/latch.rs]: No such problem, since each slot actually can handle mutiple different hashes.

Comment (zhangjinpeng1987) [reply src/storage/txn/latch.rs]: Each lock only can hold one mutex at a time.
https://github.com/tikv/tikv/pull/6094/files#diff-56bcb7f504d93784d808674ed821bdbaR161-R162

Comment (sticnarf) [reply src/storage/txn/latch.rs]: Ah, sorry. Never mind. It's ok to sort by key hash.

File: src/storage/txn/latch.rs (hunk 20 above)
Comment (sticnarf) [inline src/storage/txn/latch.rs]: required -> requires? Maybe this sentence can be written better but I don't know how either.......

File: src/storage/txn/latch.rs (hunk 5 above)
Comment (breeswish) [reply src/storage/txn/latch.rs]: Why do we need concurrent data structures? Scheduler is single threaded if I remember correctly.
